### ISSUE TITLE
[Feature Request] crawl through the onion network to prevent abuse reports

### ISSUE THREAD

**Issue by Motophan** [bitmagnet-io/bitmagnet#65]
**Have you checked the roadmap on the website, and existing issues, before opening a dupllcate issue?**
Yes

**Is your feature request related to a problem? Please describe.**
If you run this program you will start to show up in iknowwhatyoudownload.com and could potentially start receiving abuse reports. 

**Describe the solution you'd like**
Let us route everything through tor if we want to, disabled by default. Make sure dns is proxied to prevent rdns leakage. Either a docker-compose tor container or system level tor service and have this container route through it either via the socks5 proxy on localhost, or just let us have a enviroment variable that is respected, like
`SOCKS_PROXY=127.0.0.1:9050`

```
services:
  bitmagnet:
    build:
      context: .
    container_name: bitmagnet
    #     image: ghcr.io/bitmagnet-io/bitmagnet:latest
    volumes:
      - ./data/bitmagnet:/tmp
    #ports:
      #- "3333:3333"
      # BitTorrent ports, wont work
      #- "3334:3334/tcp"
      #- "3334:3334/udp"
    restart: unless-stopped
    environment:
      - LOG_LEVEL=debug
      - POSTGRES_HOST=postgres
      - POSTGRES_PASSWORD=postgres
      - REDIS_ADDR=redis:6379
      - SOCKS_PROXY=127.0.0.1:9050
    command:
      - worker
      - run
      - --keys=http_server
      - --keys=queue_server
      - --keys=dht_crawler
    depends_on:
      postgres:
        condition: service_healthy
      redis:
        condition: service_healthy
```

**Describe alternatives you've considered**
I figured out how to route this over a vpn but not over tor. What I did was 
1) create a wireguard container with mullvad
2) set this container as "network mode container" and the container the wireguard mullvad container
However you cannot expose ports. 

**Additional context**
Im not sure open ports are needed other than performance as the other end would need an open port to serve metainfo to us. If I need to run this for a few months thats fine. 

tldr; I want to run the crawler through tor.

**Comment by mgdigital** [comment]
Hi, I'm not sure Tor integration is feasible, and is likely to be prohibitively slow. See https://blog.torproject.org/bittorrent-over-tor-isnt-good-idea/ or Google "BitTorrent over Tor" to find a ton of articles like this. In any case DHT uses UDP which isn't possible over Tor so this would just be for the peer protocol bit that downloads the meta info.

I DO want to make it easier to route through something like GlueTun, this might just be a case of providing an example docker-compose people can use.

**Comment by Motophan** [comment]
Ah, sorry for the miscommunication, when I said "route everything over tor" I meant this container's public connections, not the entire download stack from bitmagnet -> arr -> client. I want to crawl dht files over tor only, however I would use a seedbox for the downloads. Crawling the network isnt that much bandwidth and the onion network is better setup for this. I think maybe 10-20KB per file, so even a million files would only be a few GBs. 

A vpn / seedbox will usually ignore a few dmcas before banning you, but after running this only a hour I was listed for over 100 files on iknowwhatyoudownload.com and I think a overzelous copy protection agencies would send a letter on a metainfo req instead of waiting for you to download from them. However, they might actually wait to get data back from you before firing off a love letter. I dont know. 


However, do you get abuse reports from running this? I may be overly concerned. 

Can I contact you over email? There is something private I want to say.

**Comment by mgdigital** [comment]
I'm not sure I understand what kind of integration you have in mind, but for any part of the crawler to run over Tor would likely incur a significant overhead and might well be viewed as abuse of the Tor network. It's more about number of requests than size - the crawler might be making thousands of requests per minute. That said, I'm no Tor expert and if someone can come up with a workable PoC I'd be happy to be proven wrong.

Sure, feel free to get in touch at hello@bitmagnet.io

**Comment by mgdigital** [comment]
There's a gluetun example here now @Motophan https://github.com/bitmagnet-io/bitmagnet/blob/main/docker-compose.yml
